### ISSUE TITLE
Can we merge YOLOV9 pre-trained and custom dataset

### ISSUE THREAD

**Issue by g7crservice-pavin-kr** [ultralytics/ultralytics#14263]
Question:
Can I combine an existing Yolo dataset with a custom dataset I created? If so, how would  we do it??

Also, please provide documents and references to proceed further as we have the scenario to use yolov9s model for object detection but due to limitations of 80 classes present in yolov 9 not all objects are detected successfully so we tried to use pre-built yolo dataset with custom dataset.

Need urgent assistance to proceed further. Additionally, I want to know for custom dataset how many images should be trained so detection will go well, as currently we are not getting good results from custom dataset training.

**Comment by g7crservice-pavin-kr** [comment]
There is no bug, but needed support for custom dataset integration with YOLOV9 pre-built dataset.

**Comment by glenn-jocher** [comment]
Hello @g7crservice-pavin-kr-kr,

Certainly! Merging datasets can be a bit intricate, but I'll guide you through it. Here’s a concise breakdown:

### Merging Datasets
1. **Combine Images and Labels**:
   - Ensure both datasets are in the YOLO format.
   - Merge the images and corresponding label files into a unified directory structure.

2. **Update the `data.yaml` File**:
   - Create a `data.yaml` file that includes paths to your combined dataset and lists all class names.
   - Example structure:
     ```yaml
     path: ../datasets/combined  # dataset root dir
     train: images/train  # train images (relative to 'path')
     val: images/val  # val images (relative to 'path')
     test: images/test  # test images (optional)

     # Classes
     names:
       0: class1
       1: class2
       ...
       N: classN
     ```

### Pre-trained YOLOv9 Model
You can download pre-trained YOLOv9 models directly from the [Ultralytics YOLOv9 documentation](https://docs.ultralytics.com/models/yolov9). These models come with the necessary weights and configuration files.

### Training with Combined Dataset
Here’s a quick example of how to train with your combined dataset:
```python
from ultralytics import YOLO

# Load a pre-trained YOLOv9 model
model = YOLO('yolov9s.pt')

# Train the model on your combined dataset
results = model.train(data='path/to/data.yaml', epochs=100, imgsz=640)
```

### Additional Resources
For more detailed instructions on dataset preparation and training, you can refer to the [Ultralytics documentation on object detection datasets](https://docs.ultralytics.com/datasets/detect/).

If you need further assistance, please ensure you are using the latest version of the Ultralytics packages. Providing a reproducible example can also help us diagnose any issues more effectively: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out with any more questions. Happy training! 🚀

**Comment by g7crservice-pavin-kr** [comment]
Thanks @glenn-jocher ,
So, the roboflow part has been created while creating custom dataset and below yaml file have the same details.
Is it mandatory to have that part in yaml file after merging custom and pre-trained dataset?

Here, I added charger as custom class in the last.

This below one is the yaml file:

```
train: ../train/images
val: ../valid/images
# test: ../test/images

nc: 81
names: ['person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck',
'boat', 'traffic light', 'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse',
'sheep', 'cow', 'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase',
'frisbee', 'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
'tennis racket', 'bottle', wine glass', 'cup', 'fork', 'knite', 'spoon', 'bowl', 'banana', 'apple', 'sandwich',
'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch', 'potted plant', 'bed',
'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'ceil phone', 'microwave', 'oven',
'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear', 'hair drier', 'toothbrush','charger']

roboflow:
  workspace: xxxxx-xxxxxxxxxx
  project: charger-xxxxxxxx
  version: 1
  license: CC BY 4.0
  url: https://universe.roboflow.com/xxxxxxxxxxx/charger-jgtm6/dataset/1

**Comment by glenn-jocher** [comment]
Hello @g7crservice-pavin-kr-kr,

Thank you for sharing your `yaml` file. It's great to see that you've successfully integrated your custom class into the dataset.

Regarding your question, the `roboflow` section in your `yaml` file is not mandatory for training your model. This section is specific to Roboflow and is used for tracking and managing datasets within their platform. If you are not using Roboflow for dataset management or if you have already downloaded and prepared your dataset locally, you can safely remove this section from your `yaml` file.

Here’s a streamlined version of your `yaml` file without the `roboflow` section:

```yaml
train: ../train/images
val: ../valid/images
# test: ../test/images

nc: 81
names: ['person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck',
'boat', 'traffic light', 'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse',
'sheep', 'cow', 'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase',
'frisbee', 'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple', 'sandwich',
'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch', 'potted plant', 'bed',
'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone', 'microwave', 'oven',
'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear', 'hair drier', 'toothbrush', 'charger']
```

### Next Steps
1. **Verify Dataset Paths**: Ensure that the paths specified in the `train` and `val` fields correctly point to your image directories.
2. **Train Your Model**: Use the updated `yaml` file to train your model with the combined dataset.

```python
from ultralytics import YOLO

# Load a pre-trained YOLOv9 model
model = YOLO('yolov9s.pt')

# Train the model on your combined dataset
results = model.train(data='path/to/your/data.yaml', epochs=100, imgsz=640)
```

If you encounter any issues, please ensure you are using the latest version of the Ultralytics packages. Providing a reproducible example can also help us diagnose any issues more effectively: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any more questions. Happy training! 🚀

**Comment by g7crservice-pavin-kr** [comment]
Thanks @glenn-jocher ,
Once training is done, what is the procedure to test the model (detect objects)?
Can you further explain this above requirement along with the code?

**Comment by g7crservice-pavin-kr** [comment]
Hi @glenn-jocher,
After the training, the custom object "Charger" was not properly detected and the results were not satisfied.
Due to limitations if GPU machine, we used epochs = 10.
Number of images used for custom dataset were 100. (Training - 80, Valid - 20)
After training, we found that our custom class "Charger" was not available in the result class list.
Kindly refer the attached images.
![Screenshot 2024-07-10 181448](https://github.com/ultralytics/ultralytics/assets/162795942/8a841800-23d4-46dc-a1a2-46bc53ab537c)
![Screenshot 2024-07-10 181506](https://github.com/ultralytics/ultralytics/assets/162795942/8c36e464-6d32-4ead-b195-d46a1ccd7b06)
![Screenshot 2024-07-10 181523](https://github.com/ultralytics/ultralytics/assets/162795942/466915fd-f516-4187-9222-360b8a114a71)
![Screenshot 2024-07-10 181548](https://github.com/ultralytics/ultralytics/assets/162795942/e0deb560-59a2-4578-a8b1-016c0319e671)


Below is the code used for training:
`from` ultralytics import YOLO

# Load a pre-trained YOLOv9 model
model = YOLO('yolov9s.pt')
model.to('cuda')

# Train the model on your combined dataset
results = model.train(data='/xxxxxxxx/data.yaml', epochs=10, `imgsz=640)`

**Comment by glenn-jocher** [comment]
Hello @g7crservice-pavin-kr-kr-kr,

Thank you for sharing the details and the screenshots. It looks like the custom class "Charger" is not being detected correctly, which could be due to a few reasons. Let's address them step-by-step:

### 1. **Insufficient Training Data and Epochs**
With only 100 images (80 for training and 20 for validation) and 10 epochs, the model might not have had enough data or training time to learn the new class effectively. Typically, more images and longer training times are required for good performance.

### 2. **Class Label Configuration**
Ensure that the "Charger" class is correctly added to your `data.yaml` file and that the total number of classes (`nc`) is updated accordingly. It looks like you've done this, but double-checking won't hurt.

### 3. **Training Configuration**
Make sure the training configuration is optimal. Sometimes, tweaking hyperparameters like learning rate, batch size, and augmentation can help.

### Suggested Steps to Improve Detection

#### Increase Training Data and Epochs
If possible, try to gather more images for your custom class and increase the number of training epochs. This will help the model learn better.

#### Verify Class Labels
Ensure that the "Charger" class is correctly labeled in your dataset and that the `data.yaml` file reflects this accurately.

#### Example Code for Training
Here’s an updated example to ensure everything is set up correctly:

```python
from ultralytics import YOLO

# Load a pre-trained YOLOv9 model
model = YOLO('yolov9s.pt')
model.to('cuda')

# Train the model on your combined dataset
results = model.train(data='/path/to/data.yaml', epochs=50, imgsz=640, batch=16)
```

### Testing the Model
After training, you can test the model using the following code:

```python
from ultralytics import YOLO

# Load the trained model
model = YOLO('path/to/best.pt')

# Run inference on a test image
results = model.predict('path/to/test/image.jpg')

# Display the results
results[0].show()
```

### Additional Resources
For more detailed guidance on training and testing, you can refer to the [Ultralytics documentation](https://docs.ultralytics.com/).

### Reproducible Example
If the issue persists, please provide a minimum reproducible example to help us diagnose the problem more effectively: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

### Latest Version
Ensure you are using the latest version of the Ultralytics packages to benefit from the latest features and bug fixes.

Feel free to reach out if you have any more questions. Happy training! 🚀

**Comment by g7crservice-pavin-kr** [comment]
Hi @glenn-jocher ,
We increased the epochs to 25.
And, we added around 300 images for the custom "Charger" dataset (Training - 280, Valid - 20).
Still after the training, "Charger" was not present in the final list.

Should we use any other models in place of yolov9 ?

**Comment by glenn-jocher** [comment]
Hello @g7crservice-pavin-kr-kr-kr,

Thank you for the update and for your efforts in increasing the dataset size and training epochs. It's unfortunate that the "Charger" class is still not being detected. Here are a few additional steps and considerations that might help:

### Verify Dataset and Labels
1. **Check Annotations**: Ensure that the annotations for the "Charger" class are correct and consistent across all images.
2. **Class Indexing**: Double-check that the "Charger" class is correctly indexed in your `data.yaml` file and that there are no typos or mismatches.

### Model Alternatives
While YOLOv9 is a powerful model, you might want to try other versions like YOLOv8 or YOLOv5, which have shown robust performance in various scenarios. Here’s how you can switch to YOLOv8 for instance:

```python
from ultralytics import YOLO

# Load a pre-trained YOLOv8 model
model = YOLO('yolov8n.pt')
model.to('cuda')

# Train the model on your combined dataset
results = model.train(data='/path/to/data.yaml', epochs=50, imgsz=640, batch=16)
```

### Hyperparameter Tuning
Consider experimenting with different hyperparameters such as learning rate, batch size, and augmentation techniques. Sometimes, fine-tuning these parameters can significantly improve model performance.

### Testing the Model
After training, ensure you test the model properly:

```python
from ultralytics import YOLO

# Load the trained model
model = YOLO('path/to/best.pt')

# Run inference on a test image
results = model.predict('path/to/test/image.jpg')

# Display the results
results[0].show()
```

### Reproducible Example
If the issue persists, providing a minimum reproducible example can greatly help in diagnosing the problem: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

### Latest Version
Ensure you are using the latest version of the Ultralytics packages to benefit from the latest features and bug fixes.

Feel free to reach out if you have any more questions. Happy training! 🚀

**Comment by g7crservice-pavin-kr** [comment]
Thanks @glenn-jocher ,
So, if I'm about to use 'YOLOv8' model, should I download weights corresponding to that model, because till now these below weights I've downloaded from GitHub Repository.


```!wget -P /kaggle/working/yolov9/weights -q https://github.com/WongKinYiu/yolov9/releases/download/v0.1/yolov9-c.pt
!wget -P /kaggle/working/yolov9/weights -q https://github.com/WongKinYiu/yolov9/releases/download/v0.1/yolov9-e.pt
!wget -P /kaggle/working/yolov9/weights -q https://github.com/WongKinYiu/yolov9/releases/download/v0.1/gelan-c.pt
!wget -P /kaggle/working/yolov9/weights -q https://github.com/WongKinYiu/yolov9/releases/download/v0.1/gelan-e.pt

**Comment by g7crservice-pavin-kr** [comment]
Hi @glenn-jocher ,
Also, for the custom dataset "Charger", i trained the dataset using "Roboflow".
So, could this be the reason for the issue of 'improper detection' of Charger, and not having the "Charger" class in final list after training.

But, once the code for training was executed, the below statement was shown in the output terminal;
**Overriding model.yaml nc=80 with nc=81**

**Comment by g7crservice-harshita-gupta** [comment]
Hello @glenn-jocher, thanks for your support so far, but we need your assistance to fix it completely as this is the part of our requirement to do custom dataset training with yolov9s pre-built model.
As this activity is having time limit to wrap it, so your help will be highly appreciated as we already invested time since last week but not getting enough solution for it.
